### PR TITLE
gmailieer: behave more like sendmail

### LINKED ISSUE #152: gmi send doesn't accept additional arguments
I'm unable to use `gmi send` with NeoNutt, as "NeoMutt expects that the specified program interprets additional arguments as recipient addresses. NeoMutt appends all recipients after adding a --delimiter (if not already present)." (from https://neomutt.org/guide/reference.html#3-333-%C2%A0sendmail)

At least `sendmail(1)` from postfix also seems to behave like this:
> sendmail [option ...] [recipient ...]

Can you add support for this into lieer?

### PR DESCRIPTION
**Pull request by flokli**
sendmail(1) allows additional recipients to be passed in the command
line. The path to the message is never passed, it's always stdin:

> sendmail [option ...] [recipient ...]

At least neomutt makes use of the `recipient` arguments (and doesn't set
-t), so lieer needs to handle this properly.

Most sendmail implementations allow passing additional recipients in additional
arguments, but there's no way to express this to the GMail API - it basically
only supports the `-t` (`--read-recipients`) mode of sendmail, without
additional recipients.

However, MUA can make use of both `-t` and additional recipients, even in a
redundant fashion (with `-t` set, providing recipients in the message AND as
recipient in the CLI).

We need to check if the MUA asks us to do something we can't do.

The following combinations are OK:

 - When `-t` is passed, we need to check for the CLI-passed recipients to be
   equal or a subset of the ones passed in the headers.

 - When `-t` is not passed, all header-passed recipients need to be provided in
   the CLI as well.

This avoids silently not sending mail to some recipients (pretending we did),
or sending mail to recipients we didn't want to send to again.

Fixes #152

### CODE REVIEW

File: lieer/gmailieer.py [hunk 1]
```diff
@@ -707,21 +708,38 @@ def send (self, args):
...
+    header_recipients = set()
+    for field_name in ("To", "Cc", "Bcc"):
+      field_values = eml.get_all(field_name, [])
+      field_addrs = map(lambda x: email.utils.parseaddr(x)[1], field_values)
+      header_recipients = header_recipients.union(field_addrs)
+
+    if args.read_recipients:
+      if not header_recipients.issuperset(cli_recipients):
+          print ("Recipients passed via CLI, but not part of message headers: {}".format(", ".join(cli_recipients.difference(header_recipients))))
+          exit (2)
```
Comment (gauteh) [inline lieer/gmailieer.py:736]: Maybe use an exception here, just to stay consistent with rest of application.

File: lieer/gmailieer.py [hunk 2]
```diff
@@ -707,21 +708,38 @@ def send (self, args):
...
+    elif not header_recipients == cli_recipients:
+      print ("Recipients passed via CLI differ from those in message headers:")
+      print (" - {}".format(", ".join(cli_recipients)))
+      print (" - {}".format(", ".join(header_recipients)))
+      exit (2)
```
Comment (gauteh) [inline lieer/gmailieer.py:741]: Same here.

File: lieer/gmailieer.py [hunk 3]
```diff
@@ -707,21 +708,38 @@ def send (self, args):
...
-      msg = sys.stdin.buffer.read()
-      fn = 'stdin'
-    else:
-      if os.path.isabs(args.message):
-        fn = args.message
-      else:
-        fn = os.path.join(self.cwd, args.message)
-      msg = open(fn, 'rb').read()
+    msg = sys.stdin.buffer.read()
+    fn = 'stdin'
```
Comment (gauteh) [inline lieer/gmailieer.py:712]: We don't need this variable anymore, it will always be stdin.

**Comment by gauteh** [comment]
Hi, I think this looks good, just a few comments. It would be good to get this tested wih emacs as well (@skangas, #139). I think that maybe the error messages will be hard to understand without context, or at least figure out how to fix them. Maybe link to the readme or some wiki?

File: lieer/gmailieer.py (hunk 3 above)
Comment (flokli) [reply lieer/gmailieer.py:712]: Indeed. dropped `fn` here, and updated the `self.vprint` further down to not print `fn` anymore.

File: lieer/gmailieer.py (hunk 2 above)
Comment (flokli) [reply lieer/gmailieer.py:741]: I now `raise argparse.ArgumentError` with the (condensed) error message.

File: lieer/gmailieer.py (hunk 1 above)
Comment (flokli) [reply lieer/gmailieer.py:736]: I now `raise argparse.ArgumentError` with the error message.

**Comment by flokli** [comment]
@gauteh I updated the error messages a bit, they now say
> "Recipients passed via sendmail(1) arguments" … vs "those in message headers"

IMHO, this should be clear enough - we can't really document how every MUA uses sendmail, and it can be altered through configuration anyways. `lieer` now just works in more of these possible combinations, to the extend of what the GMail API allows us to do without using SMTP.

Regarding emacs: My guess would be it just passed `-t`, didn't specify any additional recipients, and passed the message as stdin - otherwise I don't see how code in master could have worked at all. 

But I agree someone running an emacs setup should at least verify sending mail still works after this change.

**Comment by gauteh** [comment]
Thanks, sounds good, but I can't see any changes? Did you push?

**Comment by flokli** [comment]
:facepalm: now I did!

**Comment by gauteh** [comment]
Thanks, merged, I changed ArgumentError to ValueError since the former needed
more arguments. I tweaked the readme a bit, trying to condense it a bit
more. It is already super-long. I think this was a very necessary
change, thanks.

**Comment by skangas** [comment]
> Hi, I think this looks good, just a few comments. It would be good to get this tested wih emacs as well (@skangas, #139). I think that maybe the error messages will be hard to understand without context, or at least figure out how to fix them. Maybe link to the readme or some wiki?

I didn't have time to test it in time before it was merged. I will report back if I see any issues.

**Comment by gauteh** [comment]
Stefan Kangas writes on May 19, 2020 2:18:
>> Hi, I think this looks good, just a few comments. It would be good to get this tested wih emacs as well (@skangas, #139). I think that maybe the error messages will be hard to understand without context, or at least figure out how to fix them. Maybe link to the readme or some wiki?
>
> I didn't have time to test it in time before it was merged. I will report back if I see any issues.

Great, the safest is to add the `-t` option in emacs as well. I
updated the wiki.

**Pull request closed without merging**